downloadservice/app.py

```diff
@@ -159,10 +183,94 @@ def login(user):
     return render_template('index.html', user=user, token=token)
 
 
-def get_upstream_session():
+def certificate_validity(certificate):
+    try:
+        x509 = OpenSSL.crypto.load_certificate(
+            OpenSSL.crypto.FILETYPE_PEM, certificate)
+        asn1_time = x509.get_notAfter()
+        return datetime.strptime(asn1_time.decode(), '%Y%m%d%H%M%S%fZ')
+    except Exception:
+        raise CertificateError('invalid certificate')
+
+
+@app.route(url_prefix + '/upload-cert', methods=['POST'])
+@authenticated
+def upload_cert(user):
+    filename = user_to_path_fragment(user) + ".crt"
+    certificate_file = app.config['CTADS_CERTIFICATE_DIR'] + filename
+
+    certificate = request.json.get('certificate')
+    try:
+        if certificate and certificate_validity(certificate).date() > \
+                (date.today()+timedelta(days=1)):
```

Comment:
```suggestion
                (date.today()+timedelta(hours=12)):
```

if it's generated by the user with 1 day validity, at the time of upload it will be necessarily a little bit less than 1 day.